### PR TITLE
WIP: Windows webauthn

### PR DESCRIPTION
**Pull request by dainnilsson**
Continuation of #76.

### CODE REVIEW

File: fido2/client.py [hunk 1]
```diff
@@ -578,3 +588,148 @@ def _ctap1_get_assertion(
...
+        )
+
+        result = self.api.make_credential(
+            rp,
+            user,
+            [{"type": "public-key", "alg": alg} for alg in algos],
+            client_data,
+            timeout or 0,
+            rk,
+            WebAuthNAuthenticatorAttachment.ANY,
```
Comment (cmilam87) [inline fido2/client.py]: When the attachment of `ANY` is specified - it prompts for Windows login credentials and doesn't use the Yubikey.
image

Here it's asking for my pin to login to Windows (not the PIN of the Yubikey)

Comment (dainnilsson) [reply fido2/client.py]: Good catch I'll change it to cross-platform for now. I would like to make it specifiable by the caller in the end, still thinking about the best way to go about that.

File: fido2/win_api.py [hunk 2]
```diff
@@ -0,0 +1,730 @@
...
+                WebAuthNMakeCredentialOptions(
+                    timeout,
+                    resident_key,
+                    platform_attachment,
+                    user_verification,
+                    attestation,
+                    exclude_credentials or [],
+                )
+            ),
+            ctypes.byref(attestation_pointer),
```
Comment (cmilam87) [inline fido2/win_api.py]: Getting an error here:
```
Traceback (most recent call last):
  File "c:\users\cmilam\repos\platform\onica-sso\cli\onica_sso\yubikey.py", line 260, in make_fido2_credential
    algos=[i['alg'] for i in key_data['pubKeyCredParams']]
  File "c:\users\cmilam\.virtualenvs\onica-sso-testing-mmmw4rsz\lib\site-packages\fido2\client.py", line 682, in make_credential
    extensions,
  File "c:\users\cmilam\.virtualenvs\onica-sso-testing-mmmw4rsz\lib\site-packages\fido2\win_api.py", line 677, in make_credential
    ctypes.byref(attestation_pointer),
  File "_ctypes/callproc.c", line 933, in GetResult
OSError: [WinError -2146893809] Object already exists
```

Comment (dainnilsson) [reply fido2/win_api.py]: Looks to me like you might be passing an `exclude_credentials` list already containing a credential for the Authenticator you are trying to register. I do plan to wrap those OSErrors in WindowsClient and replace them with something a bit more user friendly...

Comment (ITProKyle) [reply fido2/win_api.py]: > Looks to me like you might be passing an exclude_credentials list already containing a credential for the Authenticator you are trying to register.

I can confirm your suspicion. The is the error I encountered in my testing for this case.

File: fido2/win_api.py [hunk 3]
```diff
@@ -0,0 +1,730 @@
...
+        ("dwFlags", DWORD),
+        ("pwszU2fAppId", PCWSTR),
+        ("pbU2fAppId", BOOL),
+        ("pCancellationId", ctypes.POINTER(GUID)),
+        ("pAllowCredentialList", ctypes.POINTER(WebAuthNCredentialList)),
+    ]
+
+    def __init__(self, timeout, attachment, user_verification_requirement, credentials):
+        self.dwVersion = get_version(self.__class__.__name__)
+        self.dwTimeoutMilliseconds = timeout
```
Comment (ITProKyle) [inline fido2/win_api.py]: can this be made to convert seconds (int of float) into milliseconds for feature parity with the other clients?

```suggestion
        self.dwTimeoutMilliseconds = int(timeout * 1000)
```

for `WebAuthNMakeCredentialOptions` as well.

**Comment by ITProKyle** [comment]
I have the latest commit to this branch implemented in an app. It is working near flawlessly. I did a 1-for-1 swap of the `Fido2Client` with the `WindowsClient` when run on Windows and only ran into one bug (left a comment).

Is there a potential ETA for the release?

**Comment by dainnilsson** [comment]
Heads up: I've pushed some backwards incompatible changes (see new entry in NEWS).

In short, the Fido2Client API was a bit too CTAP-centric to be cleanly mapped to the Windows API. I've changes it to correspond more to the WebAuthn specification, which also means it matches the Fido2Server API better. Mainly the arguments passed to make_credential and get_assertion are different, see examples/credential.py for example usage.

File: fido2/win_api.py (hunk 3 above)
Comment (dainnilsson) [reply fido2/win_api.py]: This has been resolved by going the other way, the timeout is now given in ms as per the WebAuthn spec in both Client implementations.

File: fido2/client.py (hunk 1 above)
Comment (dainnilsson) [reply fido2/client.py]: For make_credential this is now up to the caller as per the WebAuthn spec. For get_assertion this is still hard coded to ANY, but I don't believe this causes any issues in practice (please let me know if it does).

**Comment by dainnilsson** [comment]
> Is there a potential ETA for the release?

I'm fairly satisfied with this PR, and intend to merge it and create a new release in the 1-2 week time frame. Still want to do some internal testing of the changed APIs.

**Pull request merged**